### ISSUE TITLE
VS Code problemMatcher-friendly reporter

### ISSUE THREAD

**Issue by danvk** [webpro-nl/knip#438]
I'd like to have a VS Code Task for knip that has a [problemMatcher](https://code.visualstudio.com/docs/editor/tasks#_defining-a-problem-matcher). The output format of knip isn't the best for parsing. Here's an example of what I see:

```
$ pnpm lint:knip
Analyzing workspace ....
Unused devDependencies (1)
ts-node  package.json
Unused exports (3)
runFile        function  src/main.ts:10:23
runPrompt      function  src/main.ts:15:23
tokenToString  function  src/token.ts:10:17
```

Here's a `problemMatcher` that I've written that at least matches the "Unused exports" bit:

```json
{
	"type": "npm",
	"script": "lint:knip",
	"label": "npm: lint:knip",
	"detail": "knip",
	"problemMatcher": {
		"owner": "tsc",
		"fileLocation": ["relative", "${workspaceFolder}"],
		"pattern": [
			{
				"regexp": "^([^\\s].*) \\(\\d+\\)$",
				"message": 1
			},
			{
				"regexp": "([^ ]+) +([^ ]+) +([^:]+):(\\d+):(\\d+)",
				"file": 3,
				"line": 4,
				"column": 5,
				"loop": true
			}
		]
	}
}
```

This works pretty well:

<img width="670" alt="image" src="https://github.com/webpro/knip/assets/98301/14759e07-9bf3-46e3-ae98-ee1be4b037da">

But it's a bit of a bummer that I had to write these regexes and that there aren't line numbers for the unused devDependencies.

So call this a feature request for a new reporter that's compatible with one of VS Code's built-in problem matchers, for example "tsc" or "eslint".

**Comment by webpro** [comment]
> The output format of knip isn't the best for parsing.

Did you see the JSON reporter? https://knip.dev/features/reporters#json

> This works pretty well:

Nice! Great to see Knip used like this.

> there aren't line numbers for the unused devDependencies

That's a good RFC.

> So call this a feature request for a new reporter that's compatible with one of VS Code's built-in problem matchers, for example "tsc" or "eslint".

What would be the best way to go about this? Use the JSON reporter as a basis, or implement a new reporter for VS Code? Either way, I'll try to read up on the topic a bit more when I can, I'm not familiar with this part of VS Code at all.

Anybody reading this, definitely feel free to pick this up, might be a fun task.

**Comment by danvk** [comment]
> > The output format of knip isn't the best for parsing.
> 
> Did you see the JSON reporter? https://knip.dev/features/reporters#json

Yes! You'd think JSON output would be perfect for consuming from another tool, but so far as I can tell VS Code problemMatchers really want line-based output.

> > This works pretty well:
> 
> Nice! Great to see Knip used like this.
> 
> > there aren't line numbers for the unused devDependencies
> 
> That's a good RFC.

Want me to file an issue for this?

> > So call this a feature request for a new reporter that's compatible with one of VS Code's built-in problem matchers, for example "tsc" or "eslint".
> 
> What would be the best way to go about this? Use the JSON reporter as a basis, or implement a new reporter for VS Code? Either way, I'll try to read up on the topic a bit more when I can, I'm not familiar with this part of VS Code at all.
> 
> Anybody reading this, definitely feel free to pick this up, might be a fun task.

I'd suggest implementing a new reporter that outputs in the exact same format as eslint, tsc or some other tool that VS Code has a built-in matcher for.

**Comment by beaussan** [comment]
> Yes! You'd think JSON output would be perfect for consuming from another tool, but so far as I can tell VS Code problemMatchers really want line-based output.

Did you try creating a [custom reporter](https://knip.dev/features/reporters#custom-reporters) that is vscode problemMatchers friendly ?

**Comment by danvk** [comment]
> Did you try creating a [custom reporter](https://knip.dev/features/reporters#custom-reporters) that is vscode problemMatchers friendly ?

That's what this issue is proposing :) I haven't taken a crack at it myself yet.

**Comment by webpro** [comment]
Closing this issue as part of a general cleanup to keep this project sustainable and optimize my time working on it. If you think this is inappropriate or if there is no workaround and you feel stuck, feel free to [open a new issue](./new/choose). Thanks for your understanding.
